code/lib/react-dom-shim/src/preset.ts

```diff
@@ -0,0 +1,43 @@
+import type { Options, StorybookConfig } from '@storybook/types';
+import { version } from 'react-dom/package.json';
+
+export const webpackFinal = async (config: any, options: Options) => {
+  const { legacyRootApi } = await options.presets.apply<{ legacyRootApi?: boolean }>(
+    'frameworkOptions'
+  );
+
+  const isReact18 = version.startsWith('18') || version.startsWith('0.0.0');
+  const useReact17 = legacyRootApi ?? !isReact18;
+  if (useReact17) return config;
+
+  return {
+    ...config,
+    resolve: {
+      ...config.resolve,
+      alias: {
+        '@storybook/react-dom-shim': '@storybook/react-dom-shim/react-18',
+      },
+    },
+  };
+};
+
+export const viteFinal = async (config: any, options: Options) => {
+  const { legacyRootApi } = await options.presets.apply<{ legacyRootApi?: boolean }>(
+    'frameworkOptions'
+  );
+
+  const isReact18 = version.startsWith('18') || version.startsWith('0.0.0');
```

Comment:
What is the reason for the fallback for `|| version.startsWith('0.0.0')`?

Reply:
I just moved this from the existing place:

https://github.com/storybookjs/storybook/blob/217e21587415fb94d810c181a87aaa08d9b0096b/code/presets/react-webpack/src/framework-preset-react-dom-hack.ts#L15

Maybe @valentinpalkovic can tell us if it is still needed (it was [his commit](https://github.com/storybookjs/storybook/commit/0f118a38dd5cfc7c0c0b5297570aae0be4057527) originally)

Reply:
That's a very good question. I think it had something to do with prereleases?! But I can't remember correctly anymore.